### ISSUE TITLE
Unable to access a collection: Unexpected end of JSON input

### ISSUE THREAD

**Issue by xtatanx** [netlify/netlify-cms#5102]
Hello, we are currently maintaining a site using Gatsby.js and Netlify CMS. Last week the CMS started failing to load a collection(file collection) of pages returning the error: `Unexpected end of JSON input`. We are getting the error on mediaLibrary.js and entrie.js resulting in the CMs not loading that collection. This issue only happens when we run the project locally, on production the CMS is working as expected.

**Describe the bug**
1. we have a site with a collection named pages
2. we have a site with over 1000 images
3. we are using bitbucket as the backend
4. mediaLibrary.js and entries.js fails with the Unexpected end of JSON input error, hence those collections are inaccessible

**Applicable Versions:**

 - Netlify CMS version: 2.11.26
 - Git provider: Bitbucket
 - OS: Mac OS Catalina

 - Node.JS version: 12+

**CMS configuration**
`init({
    config: {
        slug: {
            clean_accents: true,
        },
        backend: {
            name: 'bitbucket',
            repo: [our-repo-name],
            branch,
            cmsLabelPrefix: branch,
            auth_type: 'implicit',
            app_id: process.env.NETLIFY_APP_ID,
        },
        media_folder: 'static/uploads',
        public_folder: '/uploads',
        publish_mode: 'editorial_workflow',
        local_backend: env === 'local',
        collections: CMS_COLLECTIONS,
    },
});`

I appreciate it if you guys could help us figure out this error, we have been trying to debug it ourselves but we haven't found what is causing those errors to throw. If you need more information I'll happily add it.

Regards

**Comment by smashercosmo** [comment]
Yes, more information is definitely needed. There was this section in the issue template, that you skipped
```
**To Reproduce**
<!--
Steps to reproduce the behavior. For example:
1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error
-->
```
please provide runnable example where this issue can be reliably reproduced.

**Comment by xtatanx** [comment]
Hello guys, we will work on creating a repo with the error so you guys can run it. Thanks!

**Comment by xtatanx** [comment]
Hello guys and sorry for the delay:
Here https://github.com/jfgamez/error-netlify-CMS you can find a repo with a reproducible error. We created a repo with over 600 images resulting in a size of more than 400mb, whenever this happens, the CMS fails to get the media.

We created a script that generated different sizes and numbers of images and identified that the issue was not given by the number of images but rather the size of the folder.

**To Reproduce**
1. Download repo and install dependencies
2. Run `yarn develop:local-cms`
3. Access the CMS admin
4. Open the console
5. Navigate to the collection of the pages and select `Home`
6. Check the API errors in the console

I expect this is useful for you guys, please let me know if there is anything else we can do to help solve this issue.

**Comment by smashercosmo** [comment]
Well, this is happening because node can't handle large json objects
https://github.com/nodejs/node/issues/35973
https://github.com/nodejs/node/issues/9489

@erezrokah why do we need to base64-encode all the images? Why can't they be just loaded from their URLs?

**Comment by erezrokah** [comment]
> @erezrokah why do we need to base64-encode all the images?

We do that in several places. Which one are you referring too? We don't keep image URLs since that doesn't play nice with private repos (some contain a token that expires after a few minuets hence the usage of `URL.createObjectURL()` to get an object URL).

**Comment by smashercosmo** [comment]
I'm talking about getMedia endpoint. Do you have any idea how this issue can be solved? When there are many images or images are big enough, the resulting JSON is huge and it exceeds node's maximum string length.

**Comment by erezrokah** [comment]
For the local proxy backend we might be able to drop the base64 encoding/decoding. Not sure for other backends if they support reading raw bytes.

As a general approach maybe something like https://github.com/discoveryjs/json-ext? That would require some more research

**Comment by xtatanx** [comment]
Hello guys,

Just to follow up, is there something we c ould help to keep this conversation going and help to fix the issue?

**Comment by erezrokah** [comment]
Hi @xtatanx, I think we have all the info we need, it's just the solution requires some effort.
Another possible approach would be to implement pagination for our local backend (this is what we have for the other backends - that is mostly dictated by the git service providers).

**Comment by smashercosmo** [comment]
Just stumbled upon this myself. This issue makes it impossible to use netlify-cms with big images. @erezrokah have you thought about how to fix it?

**Comment by erezrokah** [comment]
> Just stumbled upon this myself. This issue makes it impossible to use netlify-cms with big images. @erezrokah have you thought about how to fix it?

This requires more research and I don't think there's an easy fix for it.
If you're will to use an external service, [Cloudinary](https://www.netlifycms.org/docs/cloudinary/) should work well.

**Comment by smashercosmo** [comment]
Well, I think it's only relevant for a proxy backend. So, maybe it's not that big of a change and research?

**Comment by erezrokah** [comment]
> Well, I think it's only relevant for a proxy backend. So, maybe it's not that big of a change and research?

Oh right, so I think implementing pagination here https://github.com/netlify/netlify-cms/blob/6d801ae092a6763bd751cb82bd238c7c12b31884/packages/netlify-cms-proxy-server/src/middlewares/localFs/index.ts#L92 would be a good start.

**Comment by smashercosmo** [comment]
You said before that it might be possible to drop the base64 encoding/decoding completely

**Comment by erezrokah** [comment]
> You said before that it might be possible to drop the base64 encoding/decoding completely

Yes, but I think that without pagination it will only defer the problem.
